Add alerts for long kubelet PLEG intervals

kppullin opened the pull request:
Fire warning and critical alerts when the Kubelet's Pod Liecycle Event Generator (PLEG) interval metric exceeds configured durations.

kppullin commented on config.libsonnet:
```diff
@@ -67,6 +67,8 @@
     // allow you to overcommit if you wish.
     namespaceOvercommitFactor: 1.5,
     kubeletPodLimit: 110,
+    kubeletPlegIntervalWarningSeconds: 5,
```
Note: I'm unsure of these thresholds... they're aligned well against the clusters I have access to and would fire alerts appropriately, but I lack any insight to clusters running high number of pods (or anything else that may affect the PLEG interval).

brancz commented on alerts/system_alerts.libsonnet:
```diff
@@ -187,6 +187,32 @@ local utils = import 'utils.libsonnet';
+          {
+            alert: 'KubeletPlegIntervalHigh',
+            expr: |||
+              max(kubelet_pleg_relist_interval_microseconds{quantile="0.99"} / 1000000) by (node) > %(kubeletPlegIntervalWarningSeconds)s
```
pleg_relist_duration_secondsThese metrics are deprecated and are about to be removed. Use `pleg_relist_duration_seconds` rather.

kppullin replied:
Ah, thanks for the callout.  Haven't moved to 1.14 yet : |

brancz replied:
In that case the `release-0.1` branch is for pre-1.14 releases. I'd prefer to move ahead here with the non-deprecated metric and you can create another PR against the release-0.1 branch using the deprecated metric.

brancz commented on alerts/system_alerts.libsonnet:
```diff
@@ -187,6 +187,32 @@ local utils = import 'utils.libsonnet';
+            ||| % $._config,
+            'for': '10m',
+            labels: {
+              severity: 'critical',
```
I'm not sure anyone should be paged on this. What symtom does this incur when it happens? Pods in pending? Alerts that page should really only be based on symtoms that a user notices.

kppullin replied:
In this particular instance newly scheduled pods onto this node were in the pending state (specifically `ContainerCreating`) and hanging.  Logs on the node itself highlighted the PLEG issue, which led me to the PLEG metric.  However, digging in further, the node itself was flapping between `Ready` and `NotReady`.

What I think happened:
- Node became unhealthy and started flapping.
- During the times when the node was considered `NotReady`, liveness probes failed and pods migrated to other nodes.
- Upon a push of a deployment update, the node just so happened to be in a `Ready` state and, since all other pods had been drained, k8s saw the node as having plenty of resources to allocate and scheduled the new pods to this node.

So I think you're right - there's no need to page here and `critical` is incorrect.

And now I'm wondering is if it is better to refine/extend the `KubeNodeNotReady` alert to detect flapping nodes instead of (or in addition to) this PLEG based alert.  Would something like `changes(kube_node_status_condition{status='true',condition='Ready'}[15m]) > 1` be appropriate?

kppullin replied:
And FWIW, here's a couple screenshots of the node flapping data:

`kube_node_status_condition{status='true',condition='Ready'} == 0`:

![image](https://user-images.githubusercontent.com/369946/65713382-ecb68900-e04d-11e9-8892-a9183610b88e.png)

and `changes(kube_node_status_condition{status='true',condition='Ready'}[15m]) > 1`:

![image](https://user-images.githubusercontent.com/369946/65713415-ff30c280-e04d-11e9-8838-06f84e60b40e.png)

brancz replied:
Yes I think that's a very good alert to have too. I'd say let's add a distinct one with warning severity called `KubeNodeReadinessFlapping` or something along those lines. No strong opinion whether this or a separate PR - your call.

brancz commented on alerts/system_alerts.libsonnet:
```diff
@@ -174,6 +187,19 @@ local utils = import 'utils.libsonnet';
+          {
+            alert: 'KubeletPlegIntervalHigh',
+            expr: |||
+              max(kubelet_pleg_relist_interval_seconds{quantile="0.99"}) by (node) > %(kubeletPlegIntervalWarningSeconds)s
```
The metric is a histrogram and called `kubelet_pleg_relist_duration_seconds`. I don't believe this will work as is.

kppullin replied:
I'll find some time to get a local 1.14+ cluster running and clean this up.  Hadn't noticed that the type changed from summary to histogram to go along with the changing of the metric name.

kppullin replied:
The latest commit attempts to address the change of the metric to a histogram.  I also switched from using the `interval` to `duration`, which seems clearer to me.

kppullin commented:
Found some time to circle back to this.  Updated to switch the PLEG query from the older PLEG summary metric to newer histogram metric and added the above mentioned `flapping` alert.

As always, feedback is welcome on design of the alerts, tests, etc!  Thanks for your patience : )

kppullin commented:
> I’m so sorry I did a review a few weeks ago and forgot to submit it.

No worries.  I've pushed an update incorporating your suggestions.

brancz commented on alerts/kubelet.libsonnet:
```diff
@@ -47,6 +47,32 @@
+          {
+            alert: 'KubeletPlegDurationHigh',
+            expr: |||
+              histogram_quantile(0.99, sum(rate(kubelet_pleg_relist_duration_seconds_bucket[5m])) by (node, le)) >= 10
```
We unfortunately can't rely on the node label being available here as that's not a label on the metric but added via relabeling, but we can't enforce users to have a specific configuration. We need to do a join similar to what we do [here](https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/bdf6d50253b81f9262b4cbf47f21cab7d2cc9141/alerts/kubelet.libsonnet#L40), with the `kubelet_node_name` metric that the kubelet exposes.

brancz commented on alerts/kubelet.libsonnet:
```diff
@@ -47,6 +47,32 @@
+          {
+            alert: 'KubeletPlegDurationHigh',
+            expr: |||
+              histogram_quantile(0.99, sum(rate(kubelet_pleg_relist_duration_seconds_bucket[5m])) by (instance, le) * on(instance) group_left(node) kubelet_node_name{%(kubeletSelector)s}) >= 10
```
This seems rather expensive of a query now (we need to keep in mind some people are using this with 1000s of nodes). Could you move this to a recording rule?

kppullin replied:
Sure thing - let me hack at that.

kppullin commented on rules/kubelet.libsonnet:
```diff
@@ -0,0 +1,25 @@
+        name: 'kubelet.rules',
+        rules: [
+          {
+            record: 'node_quantile:kubelet_pleg_relist_duration_seconds:histogram_quantile',
```
Open to suggestions on naming and if it's worth calculating for the three quartiles or just the one used (0.99).

brancz commented:
Love it! :+1: :100: 

Thanks a lot for the hard work!
